File: android-beans/src/main/java/rocks/frieler/android/beans/scopes/ScopedFactoryBeanDecorator.java
package rocks.frieler.android.beans.scopes;

import java8.util.function.Function;

/**
 * Decorator for {@link ScopedFactoryBean}s to enhance their behaviour.
 *
 * @param <T> the type of bean produced
 */
public class ScopedFactoryBeanDecorator<T> implements ScopedFactoryBean<T> {
    private Function<T, T> postProcessing;

    public static <T> ScopedFactoryBeanDecorator<T> decorate(ScopedFactoryBean<T> scopedFactoryBean) {
        return new ScopedFactoryBeanDecorator<>(scopedFactoryBean);
    }

    private final ScopedFactoryBean<T> delegate;

    private ScopedFactoryBeanDecorator(ScopedFactoryBean<T> delegate) {
        this.delegate = delegate;
    }

    @Override
    public String getScope() {
        return delegate.getScope();
    }

    @Override
    public Class<T> getBeanType() {
        return delegate.getBeanType();
    }

    @Override
    public T produceBean() {
        T bean = delegate.produceBean();

        if (postProcessing != null) {
            bean = postProcessing.apply(bean);
        }

        return bean;
    }

    public ScopedFactoryBeanDecorator<T> withPostProcessing(Function<T, T> postProcessing) {
        this.postProcessing = postProcessing;
        return this;
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/scopes/ScopedFactoryBeanDecorator.kt
package rocks.frieler.android.beans.scopes

import rocks.frieler.android.beans.BeansProvider
import kotlin.reflect.KClass

fun <T : Any> ScopedFactoryBean<T>.decorate(): ScopedFactoryBeanDecorator<T> = ScopedFactoryBeanDecorator(this)

/**
 * Decorator for [ScopedFactoryBean]s to enhance their behaviour.
 *
 * @param <T> the type of bean produced
 *
 * @author Christopher Frieler
 */
class ScopedFactoryBeanDecorator<T : Any> constructor(private val delegate: ScopedFactoryBean<T>) : ScopedFactoryBean<T> {
    override val scope: String
        get() = delegate.scope

    override val beanType: KClass<T>
        get() = delegate.beanType

    private var postProcessing: ((T) -> T)? = null

    override fun produceBean(dependencies: BeansProvider): T {
        var bean = delegate.produceBean(dependencies)
        postProcessing?.let { bean = it.invoke(bean) }
        return bean
    }

    fun withPostProcessing(postProcessing: (T) -> T): ScopedFactoryBeanDecorator<T> {
        this.postProcessing = postProcessing
        return this
    }
}
